controllers/nstemplateset/namespaces.go

```diff
@@ -156,30 +159,40 @@ func (r *namespacesManager) ensureInnerNamespaceResources(logger logr.Logger, ns
-func (r *namespacesManager) delete(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+// ensureDeleted ensures that the namespaces that are owned by the user (based on the label) are deleted.
+// The method deletes only one namespace in one call.
+// It returns true if all the namespaces are gone and returns false if we should re-try:
+//     If there is no namespaces found then it returns true, nil.
+//     If there is still some namespace which is not already in terminating state then it triggers
+//        the deletion of the namespace (one namespace in one call) and returns false, nil
+//     If a namespace deletion was triggered previously but is not complete yet (namespace is in terminating state)
+//        then it updates the status of the NSTemplateSet stating that some of the namespace is still in terminating state and returns false, nil.
+// If some error happened then it returns false, error
+func (r *namespacesManager) ensureDeleted(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
 	// now, we can delete all "child" namespaces explicitly
 	username := nsTmplSet.Name
 	userNamespaces, err := fetchNamespaces(r.Client, username)
 	if err != nil {
 		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to list namespace with label owner '%s'", username)
 	}
-	// delete the first namespace which (still) exists and is not in a terminating state
-	logger.Info("checking user namepaces associated with the deleted NSTemplateSet...")
 	for _, ns := range userNamespaces {
 		if !util.IsBeingDeleted(&ns) {
-			logger.Info("deleting a user namepace associated with the deleted NSTemplateSet", "namespace", ns.Name)
+			logger.Info("deleting a user namespace associated with the deleted NSTemplateSet", "namespace", ns.Name)
 			if err := r.Client.Delete(context.TODO(), &ns); err != nil {
 				return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to delete user namespace '%s'", ns.Name)
 			}
-			return true, nil
+			return false, nil // The namespace deletion is triggered so we should stop here. When the namespace is actually deleted the reconcile will be triggered again
 		}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, &corev1.Namespace{}); err != nil {
+			if errors.IsNotFound(err) {
+				continue // This namespace is gone. Check the next one.
+			}
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to get user namespace '%s'", ns.Name)
+		}
+		// No error implies namespace is in terminating state but has not been deleted yet, update status and returns false so we will re-try when the namespace is actually deleted
+		return false, r.setStatusTerminatingFailed(nsTmplSet, fmt.Sprintf("user namespace %s deletion was triggered but is not complete yet, something could be blocking ns deletion", ns.Name))
```

Comment:
Do we really want to set the condition to the failed state when any of the namespaces is in a terminating state? I mean - it would make sense to do it after some time (eg. 1 minute) but I'm not sure if we want to do it immediately.

Reply:
in addition to that, we alraedy have a similar logic in MUR - when the UserAccount is in terminating state for more than one minute, then it starts returning an error https://github.com/codeready-toolchain/host-operator/blob/master/controllers/masteruserrecord/masteruserrecord_controller.go#L292-L299 
So, I'm not sure if we need to set the failed status in the NsTemplateSet

Reply:
its not exactly immediately, its one reconcile after the deletion of namespace was triggered. If the namespace is not deleted until the next reconcile, to me it could be an indication that something is not right since namespace deletion is fairly quick. 
Regarding your second comment, if a namespace is not deleted it would block all the owning resources, which is why was updating the failed status in NStemplateset as well. 
Would you rather have a wait time of 1min before updating the status of NSTemplateSet as well?